lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -247,6 +248,176 @@ std::unique_ptr<Pass> createLowerToHWPass() {
 } // namespace msft
 } // namespace circt
 
+namespace {
+struct PartitionPass : public PartitionBase<PartitionPass> {
+  void runOnOperation() override;
+
+  void partition(MSFTModuleOp mod);
+  void partition(DesignPartitionOp part, SmallVectorImpl<Operation *> &users);
+
+private:
+  hw::SymbolCache topLevelSyms;
+};
+} // anonymous namespace
+
+/// Fill a symbol cache with all the top level symbols.
+static void populateSymbolCache(mlir::ModuleOp mod, hw::SymbolCache &cache) {
+  for (Operation &op : mod.getBody()->getOperations()) {
+    StringAttr symName = SymbolTable::getSymbolName(&op);
+    if (!symName)
+      continue;
+    // Add the symbol to the cache.
+    cache.addDefinition(symName, &op);
+  }
+  cache.freeze();
+}
+
+void PartitionPass::runOnOperation() {
+  ModuleOp outerMod = getOperation();
+  ::populateSymbolCache(outerMod, topLevelSyms);
+
+  // TODO: sort modules by number of instantiations so we only have to run one
+  // "bubbling" pass.
+  outerMod.walk([this](MSFTModuleOp mod) { partition(mod); });
+}
+
+void PartitionPass::partition(MSFTModuleOp mod) {
+  DenseMap<SymbolRefAttr, SmallVector<Operation *, 4>> partMembers;
+  mod.walk([&partMembers](Operation *op) {
+    if (auto part = op->getAttrOfType<SymbolRefAttr>("targetDesignPartition"))
+      partMembers[part].push_back(op);
+  });
+
+  SmallVector<DesignPartitionOp, 2> parts(mod.getOps<DesignPartitionOp>());
```

Comment:
Is the motivation to copy these into a SmallVector so we can iterate over them while the IR changes? If that's the case, you should also be able to say `llvm::make_early_inc_range(mod.getOps<DesignPartitionOp>())` to get the desired behavior without an explicit copy.